### PR TITLE
service shorthand

### PR DESCRIPTION
**Pull request by ublubu**
@wlan0

Setting Kind in the koki->kube transform because it doesn't happen automatically.

Service shorthand with some tests and example yamls.

Notes:
I'm using strings wrapped as IPAddr instead of net.IP because the json/yaml ser/de doesn't work well with net.IP. It might be worth figuring out the framework for performing validation (perhaps separately from the file format/conversion implementation).

### CODE REVIEW

File: converter/converters/koki_pod_to_kube_v1_pod.go [hunk 1]
```diff
@@ -20,6 +20,7 @@ func Convert_Koki_Pod_to_Kube_v1_Pod(pod *types.PodWrapper) (*v1.Pod, error) {
 	kubePod.Name = kokiPod.Name
 	kubePod.Namespace = kokiPod.Namespace
 	kubePod.APIVersion = kokiPod.Version
+	kubePod.Kind = "Pod"
```
Comment (wlan0) [inline converter/converters/koki_pod_to_kube_v1_pod.go]: Good catch! :+1:

File: converter/converters/koki_service_to_kube_v1_service.go [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+	kubeService.Labels = kokiService.Labels
+	kubeService.Annotations = kokiService.Annotations
+
+	if len(kokiService.ExternalName) > 0 {
+		kubeService.Spec.Type = v1.ServiceTypeExternalName
+		kubeService.Spec.ExternalName = kokiService.ExternalName
+		return kubeService, nil
+	}
+
+	kubeService.Spec.Type = v1.ServiceTypeClusterIP
```
Comment (wlan0) [inline converter/converters/koki_service_to_kube_v1_service.go]: What about `NodePort` and `LoadBalancer` services?

File: converter/converters/koki_service_to_kube_v1_service.go [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+
+	kubeService.Spec.Selector = kokiService.PodLabels
+	kubeService.Spec.ClusterIP = string(kokiService.ClusterIP)
+	kubeService.Spec.ExternalIPs = revertExternalIPs(kokiService.ExternalIPs)
+	err = revertSessionAffinityInto(kokiService.ClientIPAffinity, &kubeService.Spec)
```
Comment (wlan0) [inline converter/converters/koki_service_to_kube_v1_service.go]: Why are some methods using this `revert$TypeInto` pattern while others are using the `revert$Type` pattern?

File: converter/converters/kube_v1_service_to_koki.go [hunk 4]
```diff
@@ -1,11 +1,184 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	err = convertPortsInto(kubeService.Spec.Ports, kokiService)
+	if err != nil {
+		return nil, err
+	}
+
+	if kubeService.Spec.Type == v1.ServiceTypeLoadBalancer {
```
Comment (wlan0) [inline converter/converters/kube_v1_service_to_koki.go]: Where is this value set?

Comment (wlan0) [reply converter/converters/kube_v1_service_to_koki.go]: Oh I see, you're using the Into pattern...

File: converter/converters/koki_service_to_kube_v1_service.go (hunk 2 above)
Comment (ublubu) [reply converter/converters/koki_service_to_kube_v1_service.go]: It's set to NodePort later if it finds any NodePorts in the ServicePorts.
Then if there's LoadBalancer info, it'll upgrade the type to LoadBalancer.

I'll add comments explaining the process.

File: converter/converters/koki_service_to_kube_v1_service.go (hunk 3 above)
Comment (ublubu) [reply converter/converters/koki_service_to_kube_v1_service.go]: I'll add a comment about this.

The revertX methods return the kube-native value while the revertXInto methods insert values into an existing kube-native value. I use 'Into' if the revert method affects multiple subfields at once.

**Comment by wlan0** [comment]
This mostly LGTM. What's the deal with the ``revert...Into` pattern? 

It's not easy to tell where the values are being assigned when reading that pattern...

File: converter/converters/koki_service_to_kube_v1_service.go (hunk 3 above)
Comment (wlan0) [reply converter/converters/koki_service_to_kube_v1_service.go]: I figured that's what was happening

**Comment by ublubu** [comment]
My thinking is this: I'm using `revert...Into` as a convenient way to make changes across multiple fields. It's a tradeoff between having to trust that the `revert` method doesn't do anything bad and cluttering the top-level conversion method with marshaling multiple values into multiple fields. I generally prefer to avoid code that relies on side-effects like this, but IMO the syntax favors the approach I've taken.

With that said, I'll have a look through again to make sure there's enough going on in the `into` methods that it's worth taking the shortcut.

File: converter/converters/koki_service_to_kube_v1_service.go [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+	}
+
+	if kokiPort.Protocol != "" {
+		switch kokiPort.Protocol {
+		case types.ProtocolTCP:
+			kubePort.Protocol = v1.ProtocolTCP
+		case types.ProtocolUDP:
+			kubePort.Protocol = v1.ProtocolUDP
+		default:
+			return nil, fmt.Errorf(
```
Comment (wlan0) [inline converter/converters/koki_service_to_kube_v1_service.go]: why aren't you using util.TypeValueErrorf() everywhere?

We should use consistent error messaging.

Comment (ublubu) [reply converter/converters/koki_service_to_kube_v1_service.go]: I'll go through the errors and change the ones that fit to util.TypeValueErrorf. I remember that a bunch of them didn't feel like a good fit, but I'll add an appropriate format to util for those.

**Comment by ublubu** [comment]
updated, @wlan0

**Comment by ublubu** [comment]
@wlan0, yeah it is nicer without the revertXInto functions.

**Pull request merged**